File: Binary Tree Preorder Traversal.cpp
First Try:  非递归，用栈实现
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
 //先序遍历只需要一个栈
 //先访问当前节点，然后把当前节点入栈
 //接着访问左节点并入栈直到再无左节点
 //出栈开始对右节点的访问，这个流程要烂熟于心
class Solution {
public:
    vector<int> preorderTraversal(TreeNode *root) {
        stack<TreeNode *> s;
        vector<int> v;
        v.clear();
        while(root || !s.empty()){
            while(root){
                v.push_back(root->val); 
                s.push(root);
                root = root->left;
            }
            if(!s.empty()){
                root = s.top();
                s.pop();
                root = root->right;
            }
        }
        return v;
    }
};

Second Try：
class Solution {
public:
	void preorderTraversal(TreeNode *root, vector<int> &v) {
		v.push_back(root->val);
		if(root->left != nullptr) preorderTraversal(root->left, v);
		if(root->right != nullptr) preorderTraversal(root->right, v);
	}

    vector<int> preorderTraversal(TreeNode *root) {
        vector<int> v;
        v.clear();
		if(root == nullptr) return v;
        preorderTraversal(root, v);
        return v;
    }
};

File: python/Binary Tree Preorder Traversal.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a list of integers
    def preorderTraversal(self, root):
        res = []
        d = []
        if root == None:
            return res
        d.append(root)
        while len(d) != 0:
            node = d.pop()
            res.append(node.val)
            if node.right != None:
                d.append(node.right)
            if node.left != None:
                d.append(node.left)
        return res
            